File: api/routes/balance_backlog.py
"""
Endpoint para generar propuesta de distribución semanal de un backlog.

Acepta dos modos:
1. Excel upload (multipart/form-data con field "file")
2. JSON manual (multipart/form-data con field "payload" en JSON string)

Reusa la lógica del script standalone backlog_tool/generar_propuesta.py
sin modificarlo. Solo importa las funciones necesarias.
"""
import io
import json
import sys
import tempfile
import traceback
from datetime import datetime
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, File, Form, HTTPException, UploadFile
from fastapi.responses import JSONResponse, StreamingResponse

# Hacer importable backlog_tool/
ROOT = Path(__file__).parent.parent.parent
sys.path.insert(0, str(ROOT / "backlog_tool"))

import generar_propuesta as bt  # noqa: E402

router = APIRouter()


@router.get("/api/balance-backlog/catalogo")
def get_catalogo_modelos():
    """Devuelve la lista de modelos del catálogo (para dropdown del modo manual)."""
    try:
        cat = bt.fetch_catalogo()
        modelos = []
        for num, m in cat.items():
            alts = m.get("alternativas") or []
            label = f"{num} {' '.join(alts)}".strip()
            modelos.append({
                "modelo_num": num,
                "label": label,
                "total_seg_par": m.get("total_sec_per_pair", 0),
                "robot_restringido": m.get("robot_restringido", False),
            })
        modelos.sort(key=lambda x: x["modelo_num"])
        return {"modelos": modelos}
    except Exception as e:
        raise HTTPException(500, f"Error consultando catálogo: {e}")


@router.post("/api/balance-backlog")
async def balance_backlog(
    file: Optional[UploadFile] = File(None),
    payload: Optional[str] = Form(None),
    max_modelos: int = Form(5),
    max_sem: int = Form(3),
    excluir: str = Form(""),
    fijar: str = Form(""),
    no_aislar: str = Form(""),
):
    """
    Genera propuesta optimizada del backlog.

    Provee UNO de los dos:
    - file: Excel con formato simple o legacy
    - payload: JSON con {sem_inicio, sem_fin, modelos: [{nombre, total}]}

    Retorna:
    - Excel binario (Content-Disposition: attachment)
    - O JSON con la distribución y un base64 del Excel si se usa Accept: application/json
    """
    try:
        # 1) Parsear input
        if file:
            tmp = tempfile.NamedTemporaryFile(suffix=".xlsx", delete=False)
            tmp.write(await file.read())
            tmp.close()
            try:
                modelos, semanas = bt.parse_backlog(Path(tmp.name))
            finally:
                Path(tmp.name).unlink(missing_ok=True)
        elif payload:
            data = json.loads(payload)
            sem_ini = int(data["sem_inicio"])
            sem_fin = int(data["sem_fin"])
            if sem_fin < sem_ini:
                raise HTTPException(400, "sem_fin debe ser >= sem_inicio")
            semanas = list(range(sem_ini, sem_fin + 1))
            modelos = []
            for m in data.get("modelos", []):
                nombre = str(m.get("nombre", "")).strip()
                total = int(m.get("total", 0))
                if nombre and total > 0:
                    modelos.append((nombre, total))
            if not modelos:
                raise HTTPException(400, "No hay modelos válidos en el payload")
        else:
            raise HTTPException(400, "Provee 'file' o 'payload'")

        # 2) Combinar opts
        opts = {
            "max_mod": max_modelos,
            "max_sem": max_sem,
            "excluir": set(s.strip() for s in excluir.split(",") if s.strip()),
            "no_aislar": set(s.strip() for s in no_aislar.split(",") if s.strip()),
            "fijar": {},
        }
        # parse fijar: "MODELO:S1,S2|OTRO:S3,S4"
        for f in fijar.split("|") if fijar else []:
            if ":" in f:
                k, v = f.split(":", 1)
                opts["fijar"][k.strip()] = [int(x) for x in v.split(",") if x.strip().isdigit()]

        # filtrar excluidos
        if opts["excluir"]:
            modelos = [(n, t) for n, t in modelos if not any(e in n for e in opts["excluir"])]

        # 3) Consultar Supabase
        robots = bt.fetch_robots_activos()
        catalogo = bt.fetch_catalogo()
        n_robots_compartidos = sum(
            1 for r in robots
            if not r["nombre"].lower().startswith(("maq.", "remach", "perfor", "desheb", "cabina", "m-cod"))
        )
        cap_robot_sem = max(1, n_robots_compartidos * bt.HRS_SEM_POR_ROBOT)

        # 4) Distribuir
        asig, carga, info = bt.distribuir(modelos, semanas, catalogo, cap_robot_sem, opts)

        # 5) Validar
        errores = []
        for n, t in modelos:
            s = sum(asig[n].values())
            if s != t:
                errores.append(f"{n}: suma {s} != total {t}")

        # 6) Construir resumen JSON (siempre)
        resumen = {
            "semanas": semanas,
            "capacidad_robot_sem": cap_robot_sem,
            "robots_activos": n_robots_compartidos,
            "por_semana": [
                {
                    "semana": s,
                    "pares": sum(asig[n][s] for n, _ in modelos),
                    "horas_robot": round(carga[s], 1),
                    "pct_capacidad": round(carga[s] / cap_robot_sem * 100, 1),
                    "n_modelos": sum(1 for n, _ in modelos if asig[n][s] > 0),
                }
                for s in semanas
            ],
            "por_modelo": [
                {
                    "modelo": n,
                    "total": t,
                    "distribucion": {str(s): asig[n][s] for s in semanas if asig[n][s] > 0},
                    "robot_restringido": info.get(n, {}).get("robot_restringido", False),
                    "sin_catalogo": info.get(n, {}).get("sin_catalogo", False),
                }
                for n, t in modelos
            ],
            "errores": errores,
        }

        # 7) Generar Excel
        template = ROOT / "backlog_tool" / "template_visual.xlsx"
        if not template.exists():
            raise HTTPException(500, f"Template no existe: {template}")

        out_tmp = tempfile.NamedTemporaryFile(suffix=".xlsx", delete=False)
        out_tmp.close()
        bt.escribir_excel(template, Path(out_tmp.name), asig, info, semanas)

        # nombre amigable
        now = datetime.now()
        iso_week = now.isocalendar().week
        filename = f"Backlog_OPTIMIZADO_S{iso_week:02d}_{now:%Y%m%d_%H%M}.xlsx"

        # Devolver Excel + headers con resumen JSON
        with open(out_tmp.name, "rb") as fh:
            data = fh.read()
        Path(out_tmp.name).unlink(missing_ok=True)

        return StreamingResponse(
            io.BytesIO(data),
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={
                "Content-Disposition": f'attachment; filename="{filename}"',
                "X-Resumen": json.dumps(resumen),
                "Access-Control-Expose-Headers": "Content-Disposition, X-Resumen",
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        traceback.print_exc()
        raise HTTPException(500, f"Error procesando backlog: {e}")

File: frontend/src/lib/api/balance_backlog.ts
/**
 * Cliente API para el módulo Balance Backlog.
 * Wrappea POST /api/balance-backlog (Excel upload o JSON manual).
 */
const API_URL = process.env.NEXT_PUBLIC_API_URL || ''

export interface CatalogoModelo {
  modelo_num: string
  label: string
  total_seg_par: number
  robot_restringido: boolean
  imagen_url?: string
}

export interface ResumenSemana {
  semana: number
  pares: number
  horas_robot: number
  pct_capacidad: number
  n_modelos: number
}

export interface ResumenModelo {
  modelo: string
  total: number
  distribucion: Record<string, number>
  robot_restringido: boolean
  sin_catalogo: boolean
  imagen_url?: string
}

export interface ResumenMeta {
  max_mod_pedido?: number
  max_mod_usado?: number
  max_mod_real?: number
  min_factible?: number
  ajustado?: boolean
  intentos?: number[]
  demanda_model_weeks?: number
}

export interface ResumenBacklog {
  semanas: number[]
  capacidad_robot_sem: number
  robots_activos: number
  por_semana: ResumenSemana[]
  por_modelo: ResumenModelo[]
  errores: string[]
  meta?: ResumenMeta
}

export interface BalanceBacklogResult {
  blob: Blob
  filename: string
  resumen: ResumenBacklog
}

export interface ManualPayload {
  sem_inicio: number
  sem_fin: number
  modelos: { nombre: string; total: number }[]
}

export interface BalanceOptions {
  max_modelos?: number
  max_sem?: number
  excluir?: string
  fijar?: string
  no_aislar?: string
}

export async function downloadPlantillaBacklog(): Promise<void> {
  const res = await fetch(`${API_URL}/api/balance-backlog/plantilla`)
  if (!res.ok) throw new Error(`Error ${res.status} al descargar plantilla`)
  const blob = await res.blob()
  downloadBlob(blob, 'Backlog_Plantilla.xlsx')
}

export async function fetchCatalogoBacklog(): Promise<CatalogoModelo[]> {
  const res = await fetch(`${API_URL}/api/balance-backlog/catalogo`)
  if (!res.ok) throw new Error(`Error ${res.status}: ${await res.text()}`)
  const data = await res.json()
  return data.modelos || []
}

async function postBalance(form: FormData): Promise<BalanceBacklogResult> {
  const res = await fetch(`${API_URL}/api/balance-backlog`, {
    method: 'POST',
    body: form,
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(`Error ${res.status}: ${body}`)
  }
  const blob = await res.blob()
  // Filename
  let filename = 'Backlog_OPTIMIZADO.xlsx'
  const cd = res.headers.get('Content-Disposition')
  if (cd) {
    const m = cd.match(/filename="([^"]+)"/)
    if (m) filename = m[1]
  }
  // Resumen
  let resumen: ResumenBacklog = {
    semanas: [], capacidad_robot_sem: 0, robots_activos: 0,
    por_semana: [], por_modelo: [], errores: [],
  }
  const xResumen = res.headers.get('X-Resumen')
  if (xResumen) {
    try { resumen = JSON.parse(xResumen) } catch { /* ignore */ }
  }
  return { blob, filename, resumen }
}

export async function balanceFromExcel(
  file: File,
  opts: BalanceOptions = {},
): Promise<BalanceBacklogResult> {
  const form = new FormData()
  form.append('file', file)
  if (opts.max_modelos != null) form.append('max_modelos', String(opts.max_modelos))
  if (opts.max_sem != null) form.append('max_sem', String(opts.max_sem))
  if (opts.excluir) form.append('excluir', opts.excluir)
  if (opts.fijar) form.append('fijar', opts.fijar)
  if (opts.no_aislar) form.append('no_aislar', opts.no_aislar)
  return postBalance(form)
}

export async function balanceFromManual(
  payload: ManualPayload,
  opts: BalanceOptions = {},
): Promise<BalanceBacklogResult> {
  const form = new FormData()
  form.append('payload', JSON.stringify(payload))
  if (opts.max_modelos != null) form.append('max_modelos', String(opts.max_modelos))
  if (opts.max_sem != null) form.append('max_sem', String(opts.max_sem))
  if (opts.excluir) form.append('excluir', opts.excluir)
  if (opts.fijar) form.append('fijar', opts.fijar)
  if (opts.no_aislar) form.append('no_aislar', opts.no_aislar)
  return postBalance(form)
}

export function downloadBlob(blob: Blob, filename: string) {
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  a.click()
  URL.revokeObjectURL(url)
}
